### PR TITLE
Bowtie2 goes bam_native

### PR DESCRIPTION
**Pull request by bgruening**
should replace https://github.com/galaxyproject/tools-iuc/pull/1590

### CODE REVIEW

File: tools/bowtie2/bowtie2_wrapper.xml [hunk 1]
```diff
@@ -529,6 +533,9 @@ bowtie2
             </when>
         </conditional>
         <param name="save_mapping_stats" type="boolean" checked="False" label="Save the bowtie2 mapping statistics to the history" />
+        <param name="reorder" argument="--reorder" type="boolean" truevalue="--reorder" falsevalue=""
+            label="Reorder output to reflect order of the input file"
+            help="Reorder guarantees that output SAM records are printed in an order corresponding to the order of the reads in the original input file, even when -p is set greater than 1." />
```
Comment (nsoranzo) [inline tools/bowtie2/bowtie2_wrapper.xml]: This should probably go in a conditional guarded by the `sam_opt` param, which could be moved outside of the above conditional.

File: tools/bowtie2/bowtie2_wrapper.xml [hunk 2]
```diff
@@ -529,6 +533,9 @@ bowtie2
             </when>
```
Comment (nsoranzo) [inline tools/bowtie2/bowtie2_wrapper.xml]: `sorted` should be removed from lined 532 above.

File: tools/bowtie2/bowtie2_wrapper.xml [hunk 3]
```diff
@@ -612,6 +619,9 @@ bowtie2
         </data>
         <data format="bam" name="output" label="${tool.name} on ${on_string}: aligned reads (sorted BAM)">
           <filter>analysis_type['analysis_type_selector'] == "simple" or analysis_type['sam_opt'] is False</filter>
+          <change_format>
+            <when input="reorder" value="--reorder" format="bam_native" />
```
[the remark this answers is not in the document]
Comment (bgruening) [reply tools/bowtie2/bowtie2_wrapper.xml]: Sorry, I don't see it. Whats wrong here. The entire file is t-space indented, afaik.

[the remark this answers is not in the document]
Comment (nsoranzo) [reply tools/bowtie2/bowtie2_wrapper.xml]: It's actually a mix of 2-spaces and 4-spaces indentation. I'd at least use 4 spaces for new code.

**Comment by mvdbeek** [comment]
let's not do this yet, bam_native alone is no good and provides no advantages, I'll try to get a datatype for queryname sorted files and input order sorted into 18.01

**Comment by mvdbeek** [comment]
> you do not care about ordering at all you just want to have the same ordering as in the input FASTQ files

This is a contradiction. You *do* actually care about the order, and users will manage to provide totally unordered or coordinate sorted `bam_native` files into your HiC tools. So you still need to verify the sort order in HiC tools, and do the sorting if the header doesn't match ... so you might as well just output coordinate sorted BAM. Let me know if there's a logic issue in there, but this (and the fact that we can do better with a little effort) is why I'm not 👍 on this.

What I'm doing now is to write a datatype (lets call it `BamInputSorted` for now) for this purpose that is mostly the same as BamNative, with the distinction that 
 - we won't sniff to `BamInputSorted`
 - we don't allow coordinate sorted (which is valid BamNative!), 
 - we don't allow upload of this datatype.
 - we don't provide a converter (since then we might as well convert to the more strict queryname sorted alignment)
 This allows us (and our users) to rely on `BamInputSorted` as really being what you described above.

The second datatype I'm writing is queryname sorted (`BamQuerynameSorted`, which inherits from `BamInputSorted`), that needs no explanation I guess. So when you use an aligner to produce alignments  for HiC analysis you output `BamInputSorted`.

 If a user uploads any sort of valid coordinate sorted bam file he'd rely on implicit (or explicit) conversion to BamQuerynameSorted. If a user uploads queryname sorted alignment files those are correct input as well without going though any converter (since BamQuerynameSorted inherits from BamInputSorted). If a user upload `BamNative`, i.e totally unordered input we'd go through the implicit conversion. This way when a user selects input to a HiC tool that relies on the ordering that you mentioned he will only be able to select valid inputs. Does that make sense to you ?

**Comment by bgruening** [comment]
Yes it makes sense, its what we discussed in Norway but I don't see why this is so bad.

> This is a contradiction. You do actually care about the order, and users will manage to provide totally unordered or coordinate sorted bam_native files into your HiC tools.

I guess we just understand different things under `bam_native`. For me this does not mean `totally unordered` it just means a native BAM as it is produced by the downstream tool without Galaxy modifying it. Yes, the user can mess around with it, but this can happen with every text/tabular file as well. If you can get the other data types in even better, I thought we do this for 18.05/06 ...

Thanks for working on this!

**Comment by mvdbeek** [comment]
This is good from my side !

**Pull request merged**